tests/async_queryset/tests.py

```diff
@@ -0,0 +1,210 @@
+
+    async def test_ain_bulk(self):
+        res = await SimpleModel.objects.ain_bulk()
+        self.assertEqual(
+            res,
+            {self.s1.pk: self.s1, self.s2.pk: self.s2, self.s3.pk: self.s3},
+        )
+
+        res = await SimpleModel.objects.ain_bulk([self.s2.pk])
+        self.assertEqual(res, {self.s2.pk: self.s2})
+
+        res = await SimpleModel.objects.ain_bulk([self.s2.pk], field_name="id")
+        self.assertEqual(res, {self.s2.pk: self.s2})
+
+    async def test_alatest(self):
+        instance = await SimpleModel.objects.alatest("created")
+        self.assertEqual(instance, self.s3)
+
+        instance = await SimpleModel.objects.alatest("-created")
+        self.assertEqual(instance, self.s1)
+
+    async def test_aearliest(self):
+        instance = await SimpleModel.objects.aearliest("created")
+        self.assertEqual(instance, self.s1)
+
+        instance = await SimpleModel.objects.aearliest("-created")
+        self.assertEqual(instance, self.s3)
+
+    async def test_afirst(self):
+        instance = await SimpleModel.objects.afirst()
+        self.assertEqual(instance, self.s1)
+
+        instance = await SimpleModel.objects.filter(field=4).afirst()
+        self.assertIsNone(instance)
+
+    async def test_alast(self):
+        instance = await SimpleModel.objects.alast()
+        self.assertEqual(instance, self.s3)
+
+        instance = await SimpleModel.objects.filter(field=4).alast()
+        self.assertIsNone(instance)
+
+    async def test_aaggregate(self):
+        total = await SimpleModel.objects.aaggregate(total=Sum("field"))
+        self.assertEqual(total, {"total": 6})
+
+    async def test_aexists(self):
+        check = await SimpleModel.objects.filter(field=1).aexists()
+        self.assertIs(check, True)
+
+        check = await SimpleModel.objects.filter(field=4).aexists()
+        self.assertIs(check, False)
+
+    async def test_acontains(self):
+        check = await SimpleModel.objects.acontains(self.s1)
+        self.assertIs(check, True)
+        # Unsaved instances are not allowed, so use an ID known not to exist.
+        check = await SimpleModel.objects.acontains(
+            SimpleModel(id=self.s3.id + 1, field=4)
+        )
+        self.assertIs(check, False)
+
+    async def test_aupdate(self):
+        await SimpleModel.objects.aupdate(field=99)
+        qs = [o async for o in SimpleModel.objects.all()]
+        values = [instance.field for instance in qs]
+        self.assertEqual(set(values), {99})
+
+    async def test_adelete(self):
+        await SimpleModel.objects.filter(field=2).adelete()
+        qs = [o async for o in SimpleModel.objects.all()]
+        self.assertCountEqual(qs, [self.s1, self.s3])
+
+    @skipUnlessDBFeature("supports_explaining_query_execution")
+    @async_to_sync
+    async def test_aexplain(self):
+        supported_formats = await sync_to_async(getattr)(
+            connection.features, "supported_explain_formats"
+        )
```

Comment:
This crashes on MySQL because it uses a database connection to check supported formats:
```
DatabaseWrapper objects created in a thread can only be used in that same thread.
The object with alias 'default' was created in thread id 140101338507008 and this is thread id 140102036952896.
```

<details>

```
Traceback (most recent call last):
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/db/backends/base/base.py", line 694, in temporary_connection
    with self.cursor() as cursor:
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/db/backends/base/base.py", line 323, in cursor
    return self._cursor()
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/db/backends/base/base.py", line 299, in _cursor
    self.ensure_connection()
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/db/backends/base/base.py", line 282, in ensure_connection
    self.connect()
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/db/backends/base/base.py", line 265, in connect
    self.init_connection_state()
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/db/backends/mysql/base.py", line 259, in init_connection_state
    if self.features.is_sql_auto_is_null_enabled:
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/utils/functional.py", line 57, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/db/backends/mysql/features.py", line 232, in is_sql_auto_is_null_enabled
    return self.connection.mysql_server_data["sql_auto_is_null"]
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/utils/functional.py", line 57, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/db/backends/mysql/base.py", line 399, in mysql_server_data
    with self.temporary_connection() as cursor:
  File "/usr/lib/python3.10/contextlib.py", line 135, in __enter__
    return next(self.gen)
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/db/backends/base/base.py", line 694, in temporary_connection
    with self.cursor() as cursor:
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/db/backends/base/base.py", line 323, in cursor
    return self._cursor()
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/db/backends/base/base.py", line 301, in _cursor
    return self._prepare_cursor(self.create_cursor(name))
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/db/backends/base/base.py", line 290, in _prepare_cursor
    self.validate_thread_sharing()
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/db/backends/base/base.py", line 644, in validate_thread_sharing
    raise DatabaseError(
django.db.utils.DatabaseError: DatabaseWrapper objects created in a thread can only be used in that same thread. The object with alias 'default' was created in thread id 140101338507008 and this is thread id 140102036952896.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/test/testcases.py", line 1571, in skip_wrapper
    return test_func(*args, **kwargs)
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/tests/.env/lib/python3.10/site-packages/asgiref/sync.py", line 204, in __call__
    return call_result.result()
  File "/usr/lib/python3.10/concurrent/futures/_base.py", line 439, in result
    return self.__get_result()
  File "/usr/lib/python3.10/concurrent/futures/_base.py", line 391, in __get_result
    raise self._exception
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/tests/.env/lib/python3.10/site-packages/asgiref/sync.py", line 270, in main_wrap
    result = await self.awaitable(*args, **kwargs)
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/tests/async_queryset/tests.py", line 185, in test_aexplain
    supported_formats = await sync_to_async(getattr)(
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/tests/.env/lib/python3.10/site-packages/asgiref/sync.py", line 414, in __call__
    ret = await asyncio.wait_for(future, timeout=None)
  File "/usr/lib/python3.10/asyncio/tasks.py", line 408, in wait_for
    return await fut
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/tests/.env/lib/python3.10/site-packages/asgiref/current_thread_executor.py", line 22, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/tests/.env/lib/python3.10/site-packages/asgiref/sync.py", line 455, in thread_handler
    return func(*args, **kwargs)
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/utils/functional.py", line 57, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/db/backends/mysql/features.py", line 293, in supported_explain_formats
    if not self.connection.mysql_is_mariadb and self.connection.mysql_version >= (
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/utils/functional.py", line 57, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/db/backends/mysql/base.py", line 439, in mysql_is_mariadb
    return "mariadb" in self.mysql_server_info.lower()
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/utils/functional.py", line 57, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/db/backends/mysql/base.py", line 425, in mysql_server_info
    return self.mysql_server_data["version"]
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/utils/functional.py", line 57, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/db/backends/mysql/base.py", line 399, in mysql_server_data
    with self.temporary_connection() as cursor:
  File "/usr/lib/python3.10/contextlib.py", line 135, in __enter__
    return next(self.gen)
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/db/backends/base/base.py", line 698, in temporary_connection
    self.close()
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/db/backends/base/base.py", line 349, in close
    self.validate_thread_sharing()
  File "/home/jenkins/workspace/pr-mariadb/database/mysql/label/mariadb/python/python3.10/django/db/backends/base/base.py", line 644, in validate_thread_sharing
    raise DatabaseError(
django.db.utils.DatabaseError: DatabaseWrapper objects created in a thread can only be used in that same thread. The object with alias 'default' was created in thread id 140101338507008 and this is thread id 140102036952896.
```

</details>

We hit the same issue when trying to implement an async version of `CaptureQueriesContext()`. In theory a connection should be from the same thread, we're a bit puzzled :thinking: @andrewgodwin Any clues?

Reply:
Diff reproducing on SQLite: 

```
diff --git a/tests/async/tests.py b/tests/async/tests.py
index 66eece4b97..0610b6f1b0 100644
--- a/tests/async/tests.py
+++ b/tests/async/tests.py
@@ -1,13 +1,15 @@
+import _thread
 import asyncio
 import os
 from unittest import mock
 
-from asgiref.sync import async_to_sync
+from asgiref.sync import async_to_sync, sync_to_async
 
 from django.core.cache import DEFAULT_CACHE_ALIAS, caches
 from django.core.exceptions import ImproperlyConfigured, SynchronousOnlyOperation
+from django.db import connection
 from django.http import HttpResponse
-from django.test import SimpleTestCase
+from django.test import SimpleTestCase, TestCase
 from django.utils.asyncio import async_unsafe
 from django.views.generic.base import View
 
@@ -25,13 +27,22 @@ class CacheTest(SimpleTestCase):
         self.assertIs(cache_1, cache_2)
 
 
-class DatabaseConnectionTest(SimpleTestCase):
+class DatabaseConnectionTest(TestCase):
     """A database connection cannot be used in an async context."""
 
     async def test_get_async_connection(self):
         with self.assertRaises(SynchronousOnlyOperation):
             list(SimpleModel.objects.all())
 
+    async def test_validate_thread_sharing(self):
+        def func(connection):
+            print(connection.features.supports_json_field)
+
+        await sync_to_async(func)(connection)
+
+        a_getattr = sync_to_async(getattr)
+        print(await a_getattr(connection.features, "supports_json_field"))
+
 
 class AsyncUnsafeTest(SimpleTestCase):
     """

```

The first version works, wrapping the whole feature check in `sync_to_async`. The second, wrapping `getattr` causes the error.

Reply:
Using `connection` inside `sync_to_async` correctly finds a `default` connection in `_connections` (which is a `Local` instance). However, using `connection` outside of `sync_to_async` doesn't find a `default` connection in `_connections` and creates a new one. In both cases, `_connections` is the same object according to its `ID`. :exploding_head:

Reply:
To anyone curious, it's likely a problem of when (or rather, where) `connection.features` gets evaluated.

```python
await a_getattr(connection.features, "supports_json_field")
```

is executed in the following order:

```python
foo = connection.features
bar = a_getattr(foo, "supports_json_field")
await bar
```

Evaluating `connection.features` does the following:

1. Since `connection` is a `ConnectionProxy` instance, it first finds a default connection
2. It takes that connection and returns a `DatabaseFeatures` instance bound to it

That first step is tricky because it uses the `_connections` object as a cache. `_connections` is an instance of `asgiref.local.Local`, which is thread-local by design. So the same cache instance will contain different attributes depending on where (which thread) it's accessed from.

Now, when `connection.features` is evaluated in the main thread, it returns a `DatabaseFeatures` instance bound to the connection object created in the main thread (and only safe to use within that main thread).

The last puzzle piece is `sync_to_async`, which creates a one-off worker thread to complete a synchronous function call without blocking the event loop.

The `getattr` call executed within `sync_to_async` runs in a new thread. This new thread attempts to use the features object passed from the main thread. Which, in turn, tries to use the connection object created in the main thread (to check if the connection is to a MariaDB server).

Reply:
Hey @patrys — Nice. 👍 

> The last puzzle piece is... 

I'd gotten to just before this, but was missing this, so 👯 — Thanks! 

We need to have a little think about how we structure calls to not run into this — especially if they're nested somewhere inside a context manager, as @felixxm hit for `CaptureQueriesContext` — but tomorrow... 🛏️ 

🎁